graph/src/util/jobs.rs

```diff
@@ -0,0 +1,149 @@
+        // We want tasks to start running pretty soon after server start, but
+        // also want to avoid that they all need to run at the same time. We
+        // therefore run them a small fraction of their interval from now. For
+        // a job that runs daily, we'd do the first run in about 15 minutes
+        let next_run = Instant::now() + interval / 91;
+        let task = Task {
+            job,
+            interval,
+            logger,
+            next_run,
+        };
+        self.tasks.push(task);
+    }
+
+    pub async fn start(mut self) {
+        info!(
+            self.logger,
+            "Starting job runner with {} jobs",
+            self.tasks.len()
+        );
+
+        for task in &self.tasks {
+            let next = task.next_run.saturating_duration_since(Instant::now());
+            debug!(self.logger, "Schedule for {}", task.job.name();
+                   "interval_s" => task.interval.as_secs(),
+                   "first_run_in_s" => next.as_secs());
+        }
+
+        while !self.stop.load(Ordering::SeqCst) {
+            let now = Instant::now();
+            let mut next = Instant::now() + Duration::from_secs(365 * 24 * 60 * 60);
+            for task in self.tasks.iter_mut() {
+                if task.next_run < now {
+                    // This will become obnoxious pretty quickly
+                    trace!(self.logger, "Running job"; "name" => task.job.name());
+                    // We only run one job at a time since we don't want to
+                    // deal with the same job possibly starting twice.
+                    task.job.run(&task.logger).await;
+                    task.next_run = Instant::now() + task.interval;
+                }
+                next = next.min(task.next_run);
+            }
+            let wait = next.saturating_duration_since(Instant::now());
+            tokio::time::delay_for(wait).await;
+        }
+        warn!(self.logger, "Received request to stop. Stopping runner");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::sync::{Arc, Mutex};
+    use test_store::LOGGER;
+
+    struct CounterJob {
+        count: Arc<Mutex<usize>>,
+    }
+
+    #[async_trait]
+    impl Job for CounterJob {
+        fn name(&self) -> &str {
+            "counter job"
+        }
+
+        async fn run(&self, _: &Logger) {
+            let mut count = self.count.lock().expect("Failed to lock count");
+            if *count < 10 {
+                *count += 1;
+            }
+        }
+    }
+
+    // Can't figure out what is wrong with this test: the test runner logs
+    // "Running job", but this test never sees the updates to `count`.
+    // Running graph-node, I can see that the store's VacuumDeploymentJob
+    // does get run with the configured interval, and the corresponding SQL
+    // statement turns up in the Postgres logs
+    #[ignore]
+    #[tokio::test]
+    async fn jobs_run() {
```

Comment:
Don't know exactly what's going on, but try `#[tokio::test(threaded_scheduler)]`.

Reply:
Nice. That did the trick. I updated the PR